### ISSUE TITLE
[Internal release] 1.0.2

### ISSUE THREAD

**Issue by demilich1** [demilich1/metastone#95]
1.0.2 is now available for download! As always, please report all bugs you may encounter.

**Comment by mayuso** [comment]
Thank you! I will as soon as I get home. I need revenge on that face hunter that stomped me 2 patches ago.

**Comment by webadict** [comment]
It works pretty well. I'm looking to add a few more features for v1.0.3 and fix some of the smaller bugs as well.

**Comment by demilich1** [comment]
Should we really add more features in this version though? Tomorrow will be the big Hearthstone expansion announcement and some sources are already speculating about a release this month. I would really like to get 1.0.x out of the door, so we can start we all the crazy new expansion stuff.

**Comment by mayuso** [comment]
You can work on the new expansion for 1.1.0 and still release 1.0.3 before the expansion comes out if there is a need to fix bugs.

**Comment by demilich1** [comment]
Sure, branches are an option. However, I would actually develop 1.1.0 in a branch and keep the master for 1.0.3

**Comment by mayuso** [comment]
That´s up to you, it purely subjective. :)

**Comment by webadict** [comment]
Ah, there isn't much for "feature" features. It's more broadening the current existing framework so that it allows better customization for a couple of reasons. The biggest is so that MetaStone can fit more cards as they come out without modifying the underlying structure nearly as much. The goal would be to eventually have expansions that require ONLY adding the cards to the simulator, because then the simulator has done its purpose. The second is that Hearthstone is already testing new ideas with Tavern Brawls, so I'm looking to have the ability to use at least some of those same ideas in the simulator with less restrictions. Third, I'm looking to see what customization options are available for cards. I like the idea of custom cards, and they let me see what the simulator is lacking that really shouldn't be. Sure, we might not NEED some of them, but we could in the future. It's why I added a CardCostModifier to Weapons. It's not needed, but it could be. Plus, if we can get more people using the simulator, even for things like custom cards, it expands the tester base, and hopefully finds bugs easier.

These aren't really bug fixes, so I count them only as features. They're less important, but still impact the simulator in small ways. Plus, it's something to do while looking for bugs. Mostly for fun.

**Comment by demilich1** [comment]
@webadict That's of course totally fine. I personally don't think we will ever reach a state where we just create .json files when a new expansion comes out. There will always be new keywords or mechanics, which are probably impossible to predict.

That being said, I of course welcome all additions which make MetaStone a more powerful simulator. It would be great if people use it as a tool to test custom cards and being able to implement cards which do more things than in the original game is a big plus.

However those changes are not mandatory in my opinion for the first 1.0.x release. So basically we could release 1.0.2 as a more or less stable version.(if no further bugs are found). That does not mean the next release will be 1.1.0, we could easily release 1.0.3 before that. I would just like to have a stable 1.0.x version publicly released.

**Comment by Exiahs** [comment]
Would it be possible to add an option to replay a match in the exact same draw sequence? This would allow players to learn different lines of play and how hearthstone matches work in the long run

**Comment by webadict** [comment]
Kind of. The issue is that shuffle to deck abilities and spells literally
shuffle the deck (in Hearthstone), so a deck you started out with could be
changed easily. We'd also need to change a few other backends issues with
that (I don't mind that too much. The difference between random and
non-random doesn't matter much here.)

**Comment by Exiahs** [comment]
@demilich1 I see. It would be a pretty cool feature

**Comment by webadict** [comment]
I agree, and it is doable, but only to an extent. That's why I'd be hesistant to implement it. I might consider it as an extra feature.

**Comment by demilich1** [comment]
I also agree that is would certainly be possible, however depending on the implementation details, it could get a bit messy.

First we would have to offer a central place to create random numbers. Which instantly raises the question: Do you only want to keep the draw order the same when restarting the game with the same sequence or ALL random events (i.e. Knife Juggler trigger, etc)? Because if it is just the draw order, we would have to have two kinds of random numbers: one for drawing (and shuffling) and one for other stuff.

**Comment by mayuso** [comment]
I don´t think it makes sense to replay a game with ALL random events being the same. Maybe you played turn 2 differently, making the number of shots different. 

I think the only "random" events that could be repeated in a replay scenario should be shuffles and discover mechanic draw options (The 3 cards should be the same), I can´t really think of any other, because effects like Knife Jugglers passive, or Sylvanas Deathrattle are totally dependant on the board state, so the state of the board in turn 7 of the 1st game could be (and most probably will be) totally different to the state of the board in turn 7 of the 2nd game (replay game).

I hope I explained my point correctly, as you can see, once again, my English is far from perfect.

**Comment by webadict** [comment]
It's not impossible though. You simply save a seed for each game and reload
it if you want to replay the match. It's not the best way to do things, but
it does allow replaying games, including random events that happen, so long
as all random events are taken from the same random number generator.

That makes Discover actions different if you play differently though. In
fact, one additional action could change the whole game. It's not exactly
what we'd want.

Overall, it's a maybe at best. I'll see what I can do.

**Comment by Exiahs** [comment]
Ah yes, I can see the slight complication. If its an addition feature that is fine. 

Thanks in advance

**Comment by webadict** [comment]
Haha! Looks like the Player Attributes come in handy after all! I have now created Forbidden Healing. All other Forbidden cards should be doable with the same formating. Tested this with Loatheb to make sure everything works, and it certainly does.

As for research on upcoming cards, those with "If your C'Thun has 10 or more Attack" appear to only work if C'Thun is either in your deck, hand or on the field (Not destroyed or not contained within.) This includes temporary, aura, and all other attack modifiers, i.e. Playing C'Thun and buffing it with a Flametongue Totem and an Abusive Sergeant would count.

I'll get a few of the tougher cards added as well that weren't added by @kairos4242

**Comment by webadict** [comment]
My guess is they're planning to rewrite Mistcaller, so that it and C'Thun actually gain the attack inside the deck. No clue right now, but we'll see, I suppose.

**Comment by webadict** [comment]
I've got one last PR before we should release v1.0.3. It'll use formats, a few new cards, a fair number of fixes, and and a couple of spell and card updates.

Hopefully, I can get #100 working before v1.0.3.

I'll implement a working design for C'Thun based on how I believe it works. Then we can have all cards up to now working with it.

**Comment by kairos4242** [comment]
Before releasing 1.3, shouldn't we have all currently known Old Gods cards included?

**Comment by webadict** [comment]
Er, yes. That's the goal. I'm looking to implement a working version of
C'Thun with Klaxxi and a few other cards that don't have a direct working
version yes.

**Comment by kairos4242** [comment]
Oh, okay. I read your comment as only adding a few more of the cultists and then releasing, not adding all the other cards.

**Comment by nalf3in** [comment]
Hi guys, when do you think the version 1.0.3 will be out ? I tried to compile by myself but I'm only beginning in programming ..

**Comment by kairos4242** [comment]
That's demilich1's decision, but my guess is it won't be very long now, as new decks, cards, and project structure are all done. This is just a guess, but I'd say no more than a couple weeks.

**Comment by demilich1** [comment]
Are there any known exceptions or crashes open at the moment? Otherwise I would do some testing on Monday and prepare a release for testing.

**Comment by kairos4242** [comment]
None that aren't covered by webadict's PR, to my knowledge.

**Comment by webadict** [comment]
Well, we still have the buff changes, and we can move attributes and auras
into those as well.

My thought is that buffs can be added to a buff queue on a minion, given an
id and a host, so that buffs can be associated similar to auras. Maybe even
give a buff type to associate correctly with copying minions.

I'm not sure how good that sounds, and I've been testing a few solutions.
Any ideas?

**Comment by demilich1** [comment]
Yes, the new buff system is still open (I also did some work on that, but not ready to commit yet). However I think it should be in a release after 1.0.3 (which is already quite big)

**Comment by kairos4242** [comment]
@webadict sounds good to me, that's far closer to how Blizzard implemented buffing. However, that's not urgent, like demilich1 said 1.0.3 will work just fine without it. One thing that I would like in 1.0.3 is more new decks, though. The Dreamhack decklists should give us plenty of material to work with there.

Other than that, in my opinion 1.0.3 is good to go.

**Comment by nalf3in** [comment]
Nice to know that 1.0.3 is coming soon ! Also @demilich1 I would we glad to test the beta release I you want.

**Comment by webadict** [comment]
Every release is a beta release, and you're free to test every one of them
:P

**Comment by nalf3in** [comment]
I just read that akoscz was working on a potential android release but that there was dependencies issues with the UI. Was it fixed with the project structure refactor and does that mean that we can now run it on android ? Finally I plan to learn Java to start programming on android, do you mind If I use this project to do so ?

**Comment by demilich1** [comment]
@nalf3in 

@akoscz  just split up the project into different sub-projects. This was a necessary first step for further platform specific code, but there is no actual Android version of MetaStone right now. I don't know if @akoscz is actively working on that, he can probably tell you more. But you are of course welcome to use this project for learning (or anything else really, you don't even have to ask, that is the beauty of open source)

**Comment by akoscz** [comment]
The Android UI is in very early development.  I've got a bunch of local changes I haven't pushed to my remote branch yet.  Here's the diff of what I have committed so far. https://github.com/akoscz/metastone/compare/master...akoscz:android3

One limitations that I ran into this past weekend while starting to link against the `game` module is that the Java 8 code is incompatible with the existing Android build tools.  Meaning that the only Android version that we can target is [Android N](http://developer.android.com/preview/j8-jack.html) which is only a developer preview right now.  The `game` module would require some refactoring to change some of the unsupported java 8 code to java 7 code.  I'll have to dig into this a bit more.  I vaguely recall reading somewhere that some folks have ported some of the new Java 8 libraries to Android.   Lambdas and such are not a problem since we have [Retrolambda](https://github.com/orfjackal/retrolambda) as a dependency in the `android` module already.

I wouldn't count on the Android UI to be ready any time soon.  I mostly only get to hack on this stuff on the weekend :)
